Add an arg to run-api-tests to force sharding of tests

Smackteo opened the pull request:
#### ebb0da65e6aefda022b597cbd1e8f0563c54ccfd
<pre>
Add an arg to run-api-tests to force sharding of tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=298985">https://bugs.webkit.org/show_bug.cgi?id=298985</a>
<a href="https://rdar.apple.com/160728283">rdar://160728283</a>

Reviewed by Jonathan Bedard.

This change allows run-api-tests to force sharding of tests.

* Tools/Scripts/run-api-tests:
(parse_args):
* Tools/Scripts/webkitpy/api_tests/runner.py:
(Runner._shard_tests):
(Runner.run):

Canonical link: <a href="https://commits.webkit.org/300170@main">https://commits.webkit.org/300170@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9ad3dcf256f499acd6a24f77b715cc7aa787358

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121572 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41269 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31928 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128038 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/73657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/82c9c9b4-7ba1-4616-86e3-6602e9645109) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123448 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49848 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92336 "Failed to checkout and rebase branch from PR 50830") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/73657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124524 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33511 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108895 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/73023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5945960a-d274-4b9f-b168-12e120513ea6) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/120929 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/32521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/27060 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71604 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102999 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27234 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/130903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48500 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/36887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100928 "22 flakes 35 failures") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48868 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105117 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100832 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25567 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/46245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24329 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48359 "Built successfully") | | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/47830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51177 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/49512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bb76289ba329413432eefb113e40b6955b3573c6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50830

gsnedders commented on Tools/Scripts/webkitpy/api_tests/runner.py:
```diff
@@ -112,21 +112,25 @@ def command_for_port(port, args):
             if shard_prefix not in shards:
                 shards[shard_prefix] = []
             shards[shard_prefix].append(test)
+        if len(shards) == 1 or force_shard:
```
As you wrote in the commit message:

> also if only one test suite is specified it shards those tests individually

This feels like a kinda significant behavioural change that developers might hit locally (e.g., when trying to reproduce something failing on the bots), and might want a way to disable.

I think it probably makes sense to make that as a separate change?

That said, I'm not r-'ing this in case you/Jonathan/others feel this is fine.

Smackteo replied:
> This feels like a kinda significant behavioural change that developers might hit locally (e.g., when trying to reproduce something failing on the bots), and might want a way to disable.

Shards are regrouped if they're classified in mutually_exclusive_groups so functionally if the test can run in parallel it will but if not then it won't.

JonWBedard replied:
Could we just do `if force_shard:` here? I think the only reason we might want this behavior `if len(shards)` is if we start running this logic and EWS and want to force any newly added test to shard with it's test group if it wasn't in the special "system" category.

Smackteo replied:
I can change it to `if force_shard:`

But previously you had said
>  if a suite can be shared, we should shard it if only that subsuite is specified

Removing `len(shards) ` would not allow this

gsnedders commented on Tools/Scripts/run-api-tests:
```diff
@@ -141,6 +141,8 @@ def parse_args(args):
 
         optparse.make_option('--force', action='store_true', default=False,
                              help='Run all tests, even DISABLED tests'),
+        optparse.make_option('--force-shard', action='store_true', default=False,
+                             help='Force tests to run without grouping shards'),
```
Why not copy `run-webkit-tests` and have `-f, --fully-parallel`?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8b03c96c48359860f994886eaff3727781d768f6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50830

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a9ad3dcf256f499acd6a24f77b715cc7aa787358. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50830

Smackteo committed:
Add an arg to run-api-tests to force sharding of tests
https://bugs.webkit.org/show_bug.cgi?id=298985
rdar://160728283

Reviewed by Jonathan Bedard.

This change allows run-api-tests to force sharding of tests.

* Tools/Scripts/run-api-tests:
(parse_args):
* Tools/Scripts/webkitpy/api_tests/runner.py:
(Runner._shard_tests):
(Runner.run):

Canonical link: https://commits.webkit.org/300170@main
